### PR TITLE
FEATURE: Skeleton for loading threads in a side pane

### PR DESCRIPTION
**Pull request by martin-brennan**
**EXTREMELY ROUGH NO REVIEW PLEASE**

This commit includes:

* Additions to message and channel serializers for threads
* New route and controller for a single thread
* JS route for thread pane
* Extremely basic thread pane component
* Additions to channel manager to deal with threads, and ChatThread JS model
* Changes to chat publisher and existing JS to get new thread ID when message is created

### CODE REVIEW

File: plugins/chat/app/controllers/api/chat_threads_controller.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class Chat::Api::ChatThreadsController < Chat::Api
+  def show
+    render_serialized(
```
Comment (martin-brennan) [inline plugins/chat/app/controllers/api/chat_threads_controller.rb:5]: Obviously extremely dumb, no security checks etc, just to get the idea across, will not merge in this state.

File: plugins/chat/assets/javascripts/discourse/components/chat-message.js [hunk 2]
```diff
@@ -678,8 +679,15 @@ export default Component.extend({
   },
 
   @action
-  viewReply() {
-    this.replyMessageClicked(this.message.in_reply_to);
+  viewReplyOrThread() {
+    // TODO (martin) Clean this up, hack
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.js:683]: Maybe this is fine and not a hack? Better way?

File: plugins/chat/assets/javascripts/discourse/components/chat-thread-pane.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  @service currentUser;
+  @service chat;
+  @service router;
+
+  channel = null;
+  thread = null;
+
+  @action
+  closeThread() {
+    return this.router.transitionTo("chat.channel", {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-thread-pane.js:16]: Not sure if this is correct, would be nice if this did not reload the whole channel.

File: plugins/chat/assets/javascripts/discourse/routes/chat-channel-thread.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import DiscourseRoute from "discourse/routes/discourse";
+import { inject as service } from "@ember/service";
+import { action } from "@ember/object";
+import { schedule } from "@ember/runloop";
+
+export default class ChatChannelThread extends DiscourseRoute {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/routes/chat-channel-thread.js:6]: Can I please get some feedback on this to make sure it's not doing anything silly? I don't create routes often

File: plugins/chat/assets/javascripts/discourse/services/chat-channels-manager.js [hunk 5]
```diff
@@ -30,6 +32,20 @@ export default class ChatChannelsManager extends Service {
     }
   }
 
+  // TODO (martin) Maybe we should make a ChatThreadManager as well?
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/services/chat-channels-manager.js:35]: I think you will agree with this, seems like too much mixing of concerns, maybe we should have one ChatThreadManager instance per channel in ChatChannelManager, thoughts? All the other functions I added in this file would be inclluded.

File: plugins/chat/assets/javascripts/discourse/templates/chat-channel-thread.hbs [hunk 6]
```diff
@@ -0,0 +1 @@
+<FullPageChat />
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/templates/chat-channel-thread.hbs:1]: I guess this is correct? Not sure, please double check

**Comment by martin-brennan** [comment]
@jjaffeux can you please take a look at this and let me know initial thoughts? It's based on this PR https://github.com/discourse/discourse/pull/20132. You can check this branch `feature/chat-thread-client-skeleton` out locally, and then do the following to test:

* Set enable_experimental_chat_threaded_discussions to true
* Set threading_enabled to true on whatever channel you want
* Reply to an existing message to create a thread, then when you click on the reply indicator the thread panel opens

image

**Comment by jjaffeux** [comment]
Ok so Im not answering everything in details because I think we need to take a step back and pay a little bit of debt/re-architecturing on this.

I think our target routes should be something like this:

```javascript
this.route("channel", { path: "/c/:channelTitle/:channelId" }, function () {
  this.route("from-params", { path: "/" }, function () {
    this.route("thread", { path: "/t/:threadId" });
  });
  this.route("near-message", { path: "/:messageId" });
});

this.route("channel.info", { path: "/c/:channelTitle/:channelId/info" }
this.route("channel.about", { path: "/c/:channelTitle/:channelId/about" }
this.route("channel.members", { path: "/c/:channelTitle/:channelId/members" }
this.route("channel.settings", { path: "/c/:channelTitle/:channelId/settings" }
```

The basic reason for this would be that now we can use `{{outlet}}` in the template and have it super seamless. An alternative to this is to use some channel var to show or hide thread. I think the outlet path is so much more cleaner though.

One issue we have though, is that we don't know exactly how mobile/drawer threading will look.

**Comment by martin-brennan** [comment]
@jjaffeux okay those routes do look nicer, I will just put this threading work on ice for now until those routes are changed around.

> One issue we have though, is that we don't know exactly how mobile/drawer threading will look.

I was trying not to worry about drawer for now, but you're right that mobile is a concern...presumably it will look similar to the main chat stream on mobile, but in the context of a thread, and a slightly different header...not sure if `ThreadPanel` would be required in that case or some other way of doing it. It's almost like a special version of `ChatLivePane`, maybe this part could be moved into its own component:

https://github.com/discourse/discourse/blob/1c9759af5d807795d2c6425af648a0d49ec73f94/plugins/chat/assets/javascripts/discourse/components/chat-live-pane.hbs#L60-L99

Anyway, will leave this for now until I hear more about routes.

**Comment by jjaffeux** [comment]
That should make this possible:

https://github.com/discourse/discourse/pull/20179

Will try to merge it tomorrow, but feel free to use it as a base for your work.

You will need to add {{outlet}} in chat-channel template and add the thread route as a child of chat-channel, and that should be a good starting point. Let me know any issue you encounter with this.

Note it's important to build the thread panel as a component and not a template of the chat channel thread route, because we will need to component directly in chat drawer.

**Comment by martin-brennan** [comment]
@jjaffeux thanks very much for that, will take a look at that PR today and may get time to start basing my work on it :+1: 

> Note it's important to build the thread panel as a component and not a template of the chat channel thread route, because we will need to component directly in chat drawer.

Definitely, I made a `ChatThreadPane` component in this PR, will carry across a similar idea to the next PR. Will close this one off for now :+1:

**Pull request closed without merging**